base_report_to_printer/models/printing_printer.py

```diff
@@ -0,0 +1,244 @@
+            ('error', 'Error'),
+            ('server-error', 'Server Error'),
+        ],
+        required=True,
+        readonly=True,
+        default='unknown')
+    status_message = fields.Char(readonly=True)
+    model = fields.Char(readonly=True)
+    location = fields.Char(readonly=True)
+    uri = fields.Char(string='URI', readonly=True)
+    tray_ids = fields.One2many(comodel_name='printing.tray',
+                               inverse_name='printer_id',
+                               string='Paper Sources')
+
+    @api.multi
+    def _prepare_update_from_cups(self, cups_connection, cups_printer):
+        mapping = {
+            3: 'available',
+            4: 'printing',
+            5: 'error'
+        }
+        vals = {
+            'name': cups_printer['printer-info'],
+            'model': cups_printer.get('printer-make-and-model', False),
+            'location': cups_printer.get('printer-location', False),
+            'uri': cups_printer.get('device-uri', False),
+            'status': mapping.get(cups_printer.get(
+                'printer-state'), 'unknown'),
+            'status_message': cups_printer.get('printer-state-message', ''),
+        }
+        printer_uri = cups_printer['printer-uri-supported']
+        printer_system_name = printer_uri[printer_uri.rfind('/') + 1:]
+        ppd_info = cups_connection.getPPD3(printer_system_name)
+        ppd_path = ppd_info[2]
+        if not ppd_path:
+            return vals
+
+        ppd = cups.PPD(ppd_path)
+        option = ppd.findOption('InputSlot')
+        try:
+            os.unlink(ppd_path)
+        except OSError as err:
+            # ENOENT means No such file or directory
+            # The file has already been deleted, we can continue the update
+            if err.errno != errno.ENOENT:
+                raise
+        if not option:
+            return vals
+
+        vals['tray_ids'] = []
+        cups_trays = {
+            tray_option['choice']: tray_option['text']
+            for tray_option in option.choices
+        }
+
+        # Add new trays
+        vals['tray_ids'].extend([
+            (0, 0, {'name': text, 'system_name': choice})
+            for choice, text in cups_trays.items()
+            if choice not in self.tray_ids.mapped('system_name')
+        ])
+
+        # Remove deleted trays
+        vals['tray_ids'].extend([
+            (2, tray.id)
+            for tray in self.tray_ids.filtered(
+                lambda record: record.system_name not in cups_trays.keys())
+        ])
+        return vals
+
+    @api.multi
+    def print_options(self, report=None, format=None, copies=1):
+        """ Hook to set print options """
+        options = {}
+        if format == 'raw':
+            options['raw'] = 'True'
+        if copies > 1:
+            options['copies'] = str(copies)
+        if report is not None:
+            printing_act_obj = self.env['printing.report.xml.action']
+            if report.printer_tray_id:
```

Comment:
Here you get a string for report instead of a record set, according to tests
https://travis-ci.org/OCA/report-print-send/jobs/283709521#L1157

Reply:
@yvaucher manual debug shows report is a recordset singleton of ir.actions.report.  Can you give me a little guidance fixing the test.  For now I've set it to None in the test, which is a valid value for the keyword argument as the tray functionality tests I kept in a seperate test file to cover that code.

The final error is the KeyError on printer-uri-supported - I'm not sure whether I should be modifying the test vals here, or modifying the code to allow for the missing key.  In manual testing, and prior to merge all tests passed fine, so it gives me the sense that in the real world cups always returns this key, so maybe the test vals?

Reply:
The test is wrong here, the `'pdf'` string is the `format` argument (forgot to be named), not `report`.

We didn't see the error before because the test is in the base module, and the `print_options` method in that module doesn't use the `report` argument.